crates/synthesis/src/multi_controlled/mcx.rs

```diff
@@ -0,0 +1,430 @@
+/// that make the code easier to read and that are used only for synthesis.
+trait CircuitDataForSynthesis {
+    /// Appends H to the circuit.
+    fn h(&mut self, q: u32);
+
+    /// Appends X to the circuit.
+    #[allow(dead_code)]
+    fn x(&mut self, q: u32);
+
+    /// Appends T to the circuit.
+    fn t(&mut self, q: u32);
+
+    /// Appends Tdg to the circuit.
+    fn tdg(&mut self, q: u32);
+
+    /// Appends Phase to the circuit.
+    #[allow(dead_code)]
+    fn p(&mut self, theta: f64, q: u32);
+
+    /// Appends CX to the circuit.
+    fn cx(&mut self, q1: u32, q2: u32);
+
+    /// Appends CPhase to the circuit.
+    fn cp(&mut self, theta: f64, q1: u32, q2: u32);
+
+    /// Appends CCX to the circuit.
+    fn ccx(&mut self, q1: u32, q2: u32, q3: u32) -> PyResult<()>;
+
+    /// Appends RCCX to the circuit.
+    fn rccx(&mut self, q1: u32, q2: u32, q3: u32) -> PyResult<()>;
+
+    /// Compose ``other`` into ``self``, while remapping the qubits
+    /// over which ``other`` is defined. The operations are added in-place.
+    fn compose(&mut self, other: &Self, qargs_map: &[Qubit], cargs_map: &[Clbit]) -> PyResult<()>;
+
+    /// Construct the inverse circuit
+    fn inverse(&self) -> PyResult<CircuitData>;
+}
+
+impl CircuitDataForSynthesis for CircuitData {
+    /// Appends H to the circuit.
+    #[inline]
+    fn h(&mut self, q: u32) {
+        self.push_standard_gate(StandardGate::H, &[], &[Qubit(q)]);
+    }
+
+    /// Appends X to the circuit.
+    #[inline]
+    fn x(&mut self, q: u32) {
+        self.push_standard_gate(StandardGate::X, &[], &[Qubit(q)]);
+    }
+
+    /// Appends T to the circuit.
+    #[inline]
+    fn t(&mut self, q: u32) {
+        self.push_standard_gate(StandardGate::T, &[], &[Qubit(q)]);
+    }
+
+    /// Appends Tdg to the circuit.
+    #[inline]
+    fn tdg(&mut self, q: u32) {
+        self.push_standard_gate(StandardGate::Tdg, &[], &[Qubit(q)]);
+    }
+
+    /// Appends Phase to the circuit.
+    #[inline]
+    fn p(&mut self, theta: f64, q: u32) {
+        self.push_standard_gate(StandardGate::Phase, &[Param::Float(theta)], &[Qubit(q)]);
+    }
+
+    /// Appends CX to the circuit.
+    #[inline]
+    fn cx(&mut self, q1: u32, q2: u32) {
+        self.push_standard_gate(StandardGate::CX, &[], &[Qubit(q1), Qubit(q2)]);
+    }
+
+    /// Appends CPhase to the circuit.
+    #[inline]
+    fn cp(&mut self, theta: f64, q1: u32, q2: u32) {
+        self.push_standard_gate(
+            StandardGate::CPhase,
+            &[Param::Float(theta)],
+            &[Qubit(q1), Qubit(q2)],
+        );
+    }
+
+    /// Appends the decomposition of the CCX to the circuit.
+    fn ccx(&mut self, q1: u32, q2: u32, q3: u32) -> PyResult<()> {
+        self.compose(&ccx()?, &[Qubit(q1), Qubit(q2), Qubit(q3)], &[])
+    }
+
+    /// Appends RCCX to the circuit.
+    fn rccx(&mut self, q1: u32, q2: u32, q3: u32) -> PyResult<()> {
+        self.compose(&rccx()?, &[Qubit(q1), Qubit(q2), Qubit(q3)], &[])
```

Comment:
This gives us concise syntax, but I actually meant that I expect `push_standard_gate` to be faster than creating+composing a new `CircuitData` object 🙂

Reply:
I see what you mean. However, I do prefer to have a single source of truth for the definitions of standard gates including CCX and RCCX, without having to explicitly reimplement it.